compiler/rustc_span/src/lib.rs

```diff
@@ -689,6 +689,12 @@ impl Span {
         while !outer.contains(self) {
```

Comment:
```suggestion
        while !outer.contains(self) || !self.eq_ctxt(outer) {
```
The current solution looks too tailored for the `panic`'s specific case.
Maybe we should just skip spans with different contexts during lookup in general.

Reply:
If the `eq_ctxt` requirement is determined entirely by the `find_ancestor_inside` call's context (whether the resulting span is joined/diffed with something or not), then maybe `find_ancestor_inside` should have an extra parameter for enabling it.

Reply:
Or maybe `find_ancestor_inside` should work as before, but the result's context needs to be forcibly set to `outer.ctxt()`.
I don't know what is better, it needs experimenting.

Reply:
We could have a new method `find_ancestor_inside_same_ctxt` that would call `find_ancestor_in_same_ctxt` and then also check that the spans overlap. I think that should be semantically equal to your first suggestion, because the chain of ctxts can't be cyclic? The new method would then be used instead of `find_ancestor_inside` where the span needs to be joined.

Reply:
>because the chain of ctxts can't be cyclic?

I wouldn't rely on that, from what I understand `ExpnData::call_site` (unlike `ExpnData::parent`) is generally user-settable and can be anything.
E.g. `fn macro_backtrace` is doing something like a cycle check.

Reply:
Wouldn't that cause an infinite loop in the current implementation of `find_ancestor_in_same_ctxt` and `find_ancestor_inside`? To me the logic in `macro_backtrace` looks like it's just deduplicating the backtrace for self-recursive macro calls, not doing any kind of cycle check.

Reply:
Yes, looks like macro backtrace just filters away duplicates.
But I'm still not sure that infinite loops due to `call_site` cannot be artificially constructed.

In any case, even if the logic you suggest in https://github.com/rust-lang/rust/pull/112500#discussion_r1280264769 is correct, it's highly non-obvious.
It's simpler to iterate until we find something both inside the span and having the same ctxt, with a separate function or with an extra argument to `find_ancestor_inside` doesn't matter much.